docs/fomega/lazy-machine/lazy-plutus-core.tex

```diff
@@ -0,0 +1,597 @@
+        arguments may still contain pointers into the heap.  To avoid
+        this we'd probably have to do a sort of deep forcing of the
+        arguments; again, we can't make a decision about this right now.
+      \item The strategy followed in the implementation is actually slightly different
+        because the interface to built-in functions is in flux at the moment; however
+        the effect is the same.
+    \end{itemize}
+  \item The heap is implemented as an integer counter together with a
+    Haskell \texttt{IntMap} from integer locations to closures,
+    When a new entry is inserted in the heap, the counter
+    is incremented and used as a fresh heap location, and is returned 
+    to the calling code for use in environments.
+  \item The heap grows monotonically with no garbage collection or space re-use;
+    the evaluation rules ensure that the current state of the heap is propagated 
+    through the entire evaluation process.
+    
+\end{itemize}
+
+%\newpage
+\input{lmachine.tex}
+\newpage
+
+\section{A machine providing both strict and lazy evaluation}\label{sec:strict-lazy}
+The CEK machine and the L machine are quite similar, and it should be
+relatively straightforward to combine the two to produce a machine
+supporting both strict and lazy evaluation\footnote{It's tempting to
+  refer to this as \textit{sleazy evaluation}.}.
+\\
+
+\noindent The basic idea is as follows:
+\begin{itemize}
+\item The machine would provide two forms of application:
+  \begin{itemize}
+  \item \textit{Strict application}, denoted by $[M N]$
+  \item \textit{Lazy application}, denoted by $\langle M N \rangle$
+  \end{itemize}
+  \item Machine states would consist of a stack, a heap, and an environment;
+    the environment would now contain two types of entry:
+  \begin{itemize}
+  \item \texttt{Strict} entries mapping names onto values.
+  \item \texttt{Lazy} entries, mapping names onto heap locations.
+  \end{itemize}
+  \item To evaluate $[(\lambda x.M)\, N]$, $N$ would be evaluated immediately
+    to obtain a value $V$ and the current environment would be extended to
+    contain a \texttt{Strict} entry $[x \mapsto V]$.
+  \item To evaluate $\langle (\lambda x.M)\, N \rangle$, a closure
+    containing $N$ would be stored in the heap at some new location
+    $l$ and environment would be extended to contain a \texttt{Lazy}
+    entry $[x \mapsto l]$.
+  \item Evaluation of names would then proceed in the obvious way:
+    \texttt{Strict} entries would be used immediately and \texttt{Lazy}
+    entries would be evaluated as in the L machine.
+\end{itemize}
+
+
+\noindent The obvious problem here is that if we want to perform a
+strict application $[M N]$ then $N$ might contain lazy variables; for
+example, we might have $N = \langle P\, Q \rangle$.  Presumably the solution to
+this would be to introduce an operation to force all of the lazy
+variables in $N$, analogous to Haskell's \texttt{deepseq}.
```

Comment:
This wouldn't need to be an actual exposed operation, though? It could just be implicitly called by strict application.

Of course, `deepseq`-ing things destroys your laziness, so it's unclear whether a program with "mixed" application would actually benefit much from the laziness.

Reply:
Yes, I think this would probably be hidden inside the implementation.  What worries me is that you might be evaluating away, not using much gas, and then come to a force that would suddenly cost you a huge amount.

Reply:
Yeah, for sure we lose predictable performance. But it's unclear to what degree we have that anyway. AFAIK we don't know how to get a static gas model for even strict PLC (do we?), so perhaps lazy PLC isn't that much worse?

Reply:
Regarding terminology. I think I'd view a cost model as something like an extension of the semantics which assigns costs to atomic operations and provides some means of composing these so that you can get an overall cost for a program when you run it (and you'd hope that that cost would bear some relation to the actual cost on a specific platform).  This could be something like charging one unit for every step of the evaluator, maybe with a cutomised cost for each built-in function.  This is reasonably straightforward, but the challenge is to come up with an analysis which will give you a (sound) static bound for all possible executions.  There's been quite a lot of work on this for strict languages (considerably higher-level than PLC though), but indeed things become much trickier when you've got laziness.  

I'll add a sentence about that anyway.